server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java

```diff
@@ -0,0 +1,134 @@
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public final class ParseURLFunction extends Scalar<Object, String> {
+
+    private static final String NAME = "parse_url";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNTYPED_OBJECT.getTypeSignature()
+            ),
+            ParseURLFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    public ParseURLFunction(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String url = args[0].value();
+        if (url == null) {
+            return null;
+        }
+        return parseURL(url);
+    }
+
+    private final Object parseURL(String urlText) {
+        final Map<String, Object> urlMap = new HashMap<>();
+
+        URL url = null;
+
+        try {
+            url = new URL(urlText);
+        } catch (MalformedURLException e1) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                            "unable to parse url %s",
+                                                             urlText));
+        }
+
+        urlMap.put("scheme", url.getProtocol());
+        urlMap.put("userinfo", decodeText(url.getUserInfo()));
+        urlMap.put("hostname", url.getHost());
+        urlMap.put("port", url.getPort() == -1 ? null : url.getPort());
+        urlMap.put("path", decodeText(url.getPath()));
+        urlMap.put("query", decodeText(url.getQuery()));
+        urlMap.put("parameters", parseQuery(url.getQuery()));
+        urlMap.put("fragment", decodeText(url.getRef()));
+
+        return urlMap;
+    }
+
+    private final Map<String,List<String>> parseQuery(String query) {
+        if (Strings.isNullOrEmpty(query)) {
+            return null;
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        String[] parameters = query.split("&(?!amp)");
+        for (String parameter : parameters) {
+            final int idx = parameter.indexOf("=");
+            final String key = idx > 0 ? decodeText(parameter.substring(0, idx)) : decodeText(parameter);
```

Comment:
This would not work if the query part doesn't contain a attribute-value pair separated by `=` but just some string, e.g.: `https://foo.com/some_path?bar`
This is a valid query parameter according to the RFC, see https://www.rfc-editor.org/rfc/rfc3986#section-3.4.

Reply:
What do you mean with "not work". `bar` is a parameter with `null` as value:

```sql
cr> SELECT parse_url('https://foo.com/some_path?bar')['parameters'] url;                                                                                                                                         
+-----------------+
| url             |
+-----------------+
| {"bar": [null]} |
+-----------------+
SELECT 1 row in set (0.022 sec)
```

Reply:
Ah damn, your right and I just missed reading carefully enough. Sorry!